### PR TITLE
Prevent simultaneous edits to a workflow

### LINKED ISSUE #1949: Stop simultaneous edits on a workflow
Currently 2 users with edit rights can open the workflow and try to make edits on it but they save, there are two different versions of the workflow available to be saved. Ideally we want to implement a CRDT to solve this but currently expensive so our first fix here is to disable simultaneous editing and allowing the users to know that someone is currently editing the workflow. 

- [ ] We will display a modal and will also allow them to explore the workflow in a view only mode.

### PR DESCRIPTION
**Pull request by elias-ba**
## Description

This PR allows tracking and displaying the presence status of users within the workflow canvas for a given workflow. It ensures that users can see the real-time status of their colleagues, enhancing collaboration and communication within the workflow. It also allows to sort users by the time they joined the workflow canvas and assign priorities to them. The first user to join has high priority and the others have low priority. Only users who have high priority can edit the workflow; other users can only view it.

## Validation Steps

1. Login with Multiple Sessions: Open multiple browser windows and log in with different user accounts.
2. Access the Same Workflow Canvas: Navigate to the same workflow canvas page with all logged-in sessions.
3. Check Online Users Display: Verify that all online user profiles are displayed at the top of the workflow canvas header.
4. Editing Permission Check: Ensure that only the first user who joined can edit the page.
5. Notification Banner: Verify that a banner is displayed for all other users, notifying them that they can't edit the page and identifying the active user.
6. Login with Same User Account: Log in with different sessions using the same user account.
7. Check Editing Restriction: Ensure that none of these sessions can edit the workflow canvas.

#### Expected Results:

- Presence status of users is accurately tracked and displayed in real-time.
- Only the first user who joined the workflow canvas can edit it, while others see a notification banner.
- Multiple sessions using the same user account cannot edit the workflow canvas.

## Related issue

Fixes #1949 

## Review checklist

- [x] I have performed a **self-review** of my code
- [x] I have verified that all appropriate **authorization policies** (`:owner`, `:admin`, `:editor`, `:viewer`) have been implemented and tested
- [x] If needed, I have updated the **changelog**
- [ ] Product has **QA'd** this feature

### CODE REVIEW

**Comment by christad92** [comment]
- [ ] You cannot edit this workflow because it has 2 active sessions at the moment. To enable editing, close other active sessions.
- [ ] .... You can inspect this workflow and its associated ~jobs~ steps but cannot make edits.
- [ ] When a user is being promoted to an active user, can we switch them to the latest 
- [ ] If the active user is a viewer, we should not block the next user considering that the viewer can't make edits.

File: lib/lightning/workflows.ex [hunk 1]
```diff
@@ -405,4 +405,13 @@ defmodule Lightning.Workflows do
   def jobs_ordered_subquery do
     from(j in Job, order_by: [asc: j.inserted_at])
   end
+
+  def latest?(%Workflow{lock_version: version}) do
+    query =
+      from s in Snapshot,
+        where: s.lock_version > ^version,
+        select: count(s.id)
```
Comment (stuartc) [inline lib/lightning/workflows.ex:413]: This could be improved to just consider 1 record, the existance of _anything_ higher than the current version is good enough.

Perhaps:

```elixir
from(s in Snapshot,
  where: s.lock_version > ^version,
  order_by: [desc: s.lock_version],
  limit: 1
  select: true) |> Repo.exists?
```

File: lib/lightning_web/live/workflow_live/workflow_params.ex [hunk 2]
```diff
@@ -70,28 +70,19 @@ defmodule LightningWeb.WorkflowNewLive.WorkflowParams do
...
+       }),
+       do: &Ecto.Changeset.get_embed/2
 
-  defp add_disabled_to_list(list, disabled_value) do
-    Enum.map(list, fn item -> Map.put(item, "disabled", disabled_value) end)
-  end
+  defp embed_or_assoc_getter(%Ecto.Changeset{
+         data: %Lightning.Workflows.Workflow{}
+       }),
+       do: &Ecto.Changeset.get_assoc/2
```
Comment (stuartc) [inline lib/lightning_web/live/workflow_live/workflow_params.ex:85]: Nice! :+1:

File: lib/lightning_web/live/workflow_live/edit.ex [hunk 3]
```diff
@@ -836,10 +900,20 @@ defmodule LightningWeb.WorkflowLive.Edit do
...
+      Lightning.Repo.preload(assigns.project, project_users: [:user])
+      |> Map.get(:project_users)
+      |> Enum.filter(fn pu -> pu.role == :viewer end)
+      |> Enum.map(fn pu -> pu.user.id end)
+
     {:ok,
      socket
      |> authorize()
      |> assign(
+       pid: pid,
```
Comment (stuartc) [inline lib/lightning_web/live/workflow_live/edit.ex:915]: Not sure we need to keep the `pid` in assigns right? Why not call `self()` in the functions?

File: lib/lightning_web/live/workflow_live/edit.ex [hunk 4]
```diff
@@ -1072,6 +1166,14 @@ defmodule LightningWeb.WorkflowLive.Edit do
            :error,
            "Cannot delete a node in snapshot mode, switch to latest"
          )}
+
+      :presence_low_priority ->
+        {:noreply,
+         socket
+         |> put_flash(
+           :error,
+           "Cannot delete a node in low priority mode"
```
Comment (stuartc) [inline lib/lightning_web/live/workflow_live/edit.ex:1175]: Not sure we should bring the term 'low priority' to the front end, perhaps "Cannot edit workflow while someone else is currently editing"?

@christad92 thoughts?

**Review by stuartc — CHANGES_REQUESTED** [review]
This is great @elias-ba, great work!

I've made a few comments, but your call on what to change.

File: lib/lightning/workflows.ex (hunk 1 above)
Comment (elias-ba) [reply lib/lightning/workflows.ex:413]: Fixed

File: lib/lightning_web/live/workflow_live/edit.ex (hunk 4 above)
Comment (elias-ba) [reply lib/lightning_web/live/workflow_live/edit.ex]: Done

File: lib/lightning_web/live/workflow_live/edit.ex [hunk 5]
```diff
@@ -1426,8 +1570,119 @@ defmodule LightningWeb.WorkflowLive.Edit do
...
+      current_user_presence.user.id == prior_user_presence.user.id &&
+        current_user_presence.active_sessions <= 1
+
+    %{
+      presences: presences,
+      prior_user_presence: prior_user_presence,
+      current_user_presence: current_user_presence,
+      has_presence_edit_priority: has_presence_edit_priority
+    }
+  end
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply lib/lightning_web/live/workflow_live/edit.ex]: Thanks for the suggestions, this has been moved to the Presence.ex module

File: lib/lightning_web/live/workflow_live/edit.ex (hunk 3 above)
Comment (elias-ba) [reply lib/lightning_web/live/workflow_live/edit.ex]: Fixed

**Review by midigofrank — COMMENTED** [review]
Good stuff @elias-ba 🙌

File: lib/lightning/workflows.ex [hunk 6]
```diff
@@ -405,4 +405,13 @@ defmodule Lightning.Workflows do
   def jobs_ordered_subquery do
     from(j in Job, order_by: [asc: j.inserted_at])
   end
+
+  def has_newer_version?(%Workflow{lock_version: version}) do
+    from(s in Snapshot,
+      where: s.lock_version > ^version,
+      limit: 1,
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply lib/lightning/workflows.ex:412]: You are completely right, Mr. Midigo ! thanks for the catch.

File: lib/lightning/workflows.ex [hunk 7]
```diff
@@ -405,4 +405,8 @@ defmodule Lightning.Workflows do
   def jobs_ordered_subquery do
     from(j in Job, order_by: [asc: j.inserted_at])
   end
+
+  def has_newer_version?(%Workflow{lock_version: version}) do
+    from(s in Snapshot, where: s.lock_version > ^version) |> Repo.exists?()
+  end
```
Comment (stuartc) [inline lib/lightning/workflows.ex:411]: @elias-ba I'm committing a change here, this query will return true is _any_ snapshot exists with a greater version...

Which at least from my crude testing appears to have the potential to cause an infinite loop. If not that then a bunch of extra calls to reload the workflow only to get the same one back.

I'm a little confused on how to test this behaviour, I had to jam a `insert(:snapshot, lock_version: 10)` inside a test to find out that the extra calls, and if I make the function always return true then the tests take forever (assuming infinite loop here).

Basically, I think in a separate PR drop some comments on `maybe_switch_workflow_version` and `commit_latest_version` (or on slack..).

I can't quite work out exactly how this _should_ work,

**Review by stuartc — APPROVED** [review]

File: lib/lightning/workflows.ex [hunk 8]
```diff
@@ -406,7 +406,8 @@ defmodule Lightning.Workflows do
     from(j in Job, order_by: [asc: j.inserted_at])
   end
 
-  def has_newer_version?(%Workflow{lock_version: version}) do
-    from(s in Snapshot, where: s.lock_version > ^version) |> Repo.exists?()
+  def has_newer_version?(%Workflow{lock_version: version, id: id}) do
+    from(w in Workflow, where: w.lock_version > ^version and w.id == ^id)
```
Comment (elias-ba) [inline lib/lightning/workflows.ex:410]: This looks great @stuartc and way more optimal ! Thanks

**Review by christad92 — CHANGES_REQUESTED** [review]
@Elias when a user is logged in on two different windows, the message should be:  `You can't edit this workflow right now because you have 2 active sessions. Please close the other sessions to enable editing.` 

If we have this done. I am happy with to approve this

**Comment by taylordowns2000** [comment]
> You can't edit this workflow right now because you have <<number>> active sessions. Please close the other sessions to enable editing.

@christad92 what does the current text say? i'll find/replace and get this merged

**Comment by christad92** [comment]
> > You can't edit this workflow right now because you have <> active sessions. Please close the other sessions to enable editing.
> 
> @christad92 what does the current text say? i'll find/replace and get this merged.
> 
> Edit: is it this one?
> 
> `"You cannot edit this workflow because it has #{current_user_presence.active_sessions} active sessions at the moment. To enable editing, close other active sessions."`

Yes Taylor. That is it!

**Review by christad92 — APPROVED** [review]
